File: src/main/java/com/github/claudecodegui/util/PathUtils.java
package com.github.claudecodegui.util;

import com.github.claudecodegui.model.PathCheckResult;

import java.io.File;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Path utility class.
 * Provides cross-platform path handling methods including path normalization,
 * path validation, and temporary directory detection.
 */
public class PathUtils {

    // Windows path length limit
    private static final int WINDOWS_MAX_PATH = 260;
    private static final int SAFE_PATH_LENGTH = 200; // Leave room for file names

    // ==================== Path Normalization ====================

    /**
     * Convert a path into a safe filename/identifier.
     * Consistent with Claude Code's sanitizedPath logic.
     * Replaces all non-alphanumeric characters with hyphens.
     *
     * @param path the original path
     * @return the sanitized path string (non-alphanumeric characters replaced with -)
     */
    public static String sanitizePath(String path) {
        if (path == null || path.isEmpty()) {
            return "";
        }
        // Consistent with Claude Code: replace all non-alphanumeric characters with -
        // e.g., D:\Projects\MyProject becomes D--Projects-MyProject
        return path.replaceAll("[^a-zA-Z0-9]", "-");
    }

    /**
     * Resolve a path to an absolute, normalized form, collapsing {@code .} and
     * {@code ..} segments without resolving symlinks.
     *
     * <p>This mirrors Node's {@code path.resolve()} (used by the ai-bridge when it
     * decides the working directory), so the {@code ~/.claude/projects/<key>}
     * directory derived on the Java side matches the one the Claude SDK writes to.
     * {@code getCanonicalPath()} is deliberately avoided: it resolves
     * symlinks/junctions and may change case, which would re-introduce divergence.
     *
     * @param path the original path
     * @return the absolute, normalized path; or the input unchanged when null/blank
     *         or when normalization fails
     */
    public static String normalizeAbsolute(String path) {
        if (path == null || path.isEmpty()) {
            return path;
        }
        // WSL UNC paths (\\wsl.localhost\..., \\wsl$\..., and the forward-slash
        // //wsl.localhost/... form IntelliJ's getBasePath() returns on Windows) must NOT be run
        // through Paths.get(...).toAbsolutePath(): the leading "//" collapses and the path resolves
        // drive-relative (C:\wsl.localhost\...), producing a different ~/.claude/projects/<key>
        // than the SDK writes to. See WslPathUtil.resolveHomeForFileOps for the same hazard.
        if (isWslUncPath(path)) {
            return normalizeWslUncLexically(path);
        }
        try {
            return java.nio.file.Paths.get(path).toAbsolutePath().normalize().toString();
        } catch (Exception e) {
            return path;
        }
    }

    /**
     * Guard a provider daemon's requested working directory against the project
     * base so it cannot be pointed outside the project.
     *
     * <p>Returns {@code null} when there is no project base to guard against, in
     * which case the caller should keep the original cwd. When the cwd is missing
     * or resolves outside the project base, the project base is returned (clamping
     * the daemon back inside the project). When the cwd already resolves inside
     * (or equal to) the project base, the cwd is returned verbatim so legitimate
     * sub-directory selections are preserved. Comparison is done on normalized
     * absolute paths (so trailing-slash / relative differences don't bypass the
     * guard), but the returned value keeps the original path form.
     */
    public static String guardWorkingDirectory(String cwd, String projectBase) {
        if (projectBase == null || projectBase.isEmpty()) {
            return null;
        }
        String base = normalizeAbsolute(projectBase);
        if (base == null || base.isEmpty()) {
            return null;
        }
        if (!isValidWorkingDirectory(cwd)) {
            return projectBase;
        }
        String normalizedCwd = normalizeAbsolute(cwd);
        if (normalizedCwd == null || !isWithinOrEqual(normalizedCwd, base)) {
            return projectBase;
        }
        return cwd;
    }

    /** Non-null, non-empty, and not one of the sentinel strings the webview sends for "no cwd". */
    private static boolean isValidWorkingDirectory(String cwd) {
        return cwd != null && !cwd.isEmpty() && !"undefined".equals(cwd) && !"null".equals(cwd);
    }

    /** True when {@code path} equals {@code base} or is nested directly under it. */
    private static boolean isWithinOrEqual(String path, String base) {
        if (path.equals(base)) {
            return true;
        }
        String prefix = base.endsWith(File.separator) ? base : base + File.separator;
        return path.startsWith(prefix);
    }

    /** True for the WSL UNC roots ({@code \\wsl.localhost\...}, {@code \\wsl$\...}) in either slash form. */
    private static boolean isWslUncPath(String path) {
        String p = path.replace('\\', '/');
        return p.startsWith("//wsl.localhost/") || p.startsWith("//wsl$/")
                || p.equals("//wsl.localhost") || p.equals("//wsl$");
    }

    /**
     * Collapse {@code .}/{@code ..} segments lexically while preserving the leading {@code //}
     * UNC prefix and using forward slashes (which {@link WslPathUtil} accepts on both sides).
     * Used instead of {@link java.nio.file.Path#normalize()} because Java's Windows path parser
     * corrupts the {@code //wsl...} prefix (see {@link #normalizeAbsolute}).
     */
    private static String normalizeWslUncLexically(String path) {
        String p = path.replace('\\', '/');
        String[] segments = p.substring(2).split("/");
        java.util.Deque<String> stack = new java.util.ArrayDeque<>();
        for (String seg : segments) {
            if (seg.isEmpty() || ".".equals(seg)) {
                continue;
            }
            if ("..".equals(seg)) {
                if (!stack.isEmpty()) {
                    stack.pollLast();
                }
                continue;
            }
            stack.addLast(seg);
        }
        return "//" + String.join("/", stack);
    }

    /**
     * Normalize a path to Unix style (for internal storage and comparison).
     * Converts Windows backslashes to forward slashes.
     *
     * @param path the original path
     * @return the path using forward slashes
     */
    public static String normalizeToUnix(String path) {
        if (path == null || path.isEmpty()) {
            return "";
        }
        return path.replace("\\", "/");
    }

    /**
     * Normalize a path to the current platform's style.
     *
     * @param path the original path
     * @return the path using the current platform's separator
     */
    public static String normalizeToPlatform(String path) {
        if (path == null || path.isEmpty()) {
            return "";
        }
        if (PlatformUtils.isWindows()) {
            return path.replace("/", "\\");
        } else {
            return path.replace("\\", "/");
        }
    }

    /**
     * Check whether a path is in Windows format (contains a drive letter).
     *
     * @param path the path to check
     * @return true if the path is in Windows format
     */
    public static boolean isWindowsPath(String path) {
        if (path == null || path.isEmpty()) {
            return false;
        }
        // Check if it starts with a drive letter (e.g., C:, D:)
        return path.matches("^[a-zA-Z]:.*");
    }

    /**
     * Check whether a path is a UNC path (network path).
     *
     * @param path the path to check
     * @return true if the path is a UNC path
     */
    public static boolean isUncPath(String path) {
        if (path == null || path.isEmpty()) {
            return false;
        }
        return path.startsWith("\\\\") || path.startsWith("//");
    }

    /**
     * Convert an MSYS2/Git Bash-style path to a Windows path.
     * Handles the following path mappings:
     * <ul>
     *   <li>{@code /c/Users/test} → {@code C:\Users\test} (drive letter mapping)</li>
     *   <li>{@code /home/user} or {@code ~} → {@code %USERPROFILE%} (home directory)</li>
     *   <li>{@code /tmp} → {@code %TEMP%} (temporary directory)</li>
     *   <li>{@code /dev/null} → {@code NUL} (null device)</li>
     * </ul>
     * Returns the original path if conversion is not needed.
     *
     * @param path the original path
     * @return the converted Windows path, or the original path
     */
    public static String convertMsysToWindowsPath(String path) {
        if (path == null || path.isEmpty() || !PlatformUtils.isWindows()) {
            return path;
        }

        if (isWindowsPath(path) || isUncPath(path)) {
            return path;
        }

        // Handle tilde (home directory shorthand)
        if (path.equals("~") || path.startsWith("~/")) {
            String home = PlatformUtils.getHomeDirectory();
            if (home != null && !home.isEmpty()) {
                String expanded = path.equals("~") ? home : home + path.substring(1);
                return normalizeToPlatform(expanded);
            }
        }

        // All remaining conversions require a leading slash
        if (!path.startsWith("/")) {
            return path;
        }

        // /dev/null → NUL
        if (path.equals("/dev/null")) {
            return "NUL";
        }

        // Drive letter mapping: /c/... → C:\...
        if (path.matches("^/[a-zA-Z](/.*)?$")) {
            char driveLetter = Character.toUpperCase(path.charAt(1));
            String remainder = path.length() > 2 ? path.substring(2) : "";
            String windowsPath = driveLetter + ":" + (remainder.isEmpty() ? "/" : remainder);
            return normalizeToPlatform(windowsPath);
        }

        // /home/<username>/... → %USERPROFILE%/... (Git Bash maps /home/<user> to Windows home)
        if (path.startsWith("/home/")) {
            String home = PlatformUtils.getHomeDirectory();
            if (home != null && !home.isEmpty()) {
                String afterHome = path.substring("/home/".length());
                int slashIdx = afterHome.indexOf('/');
                if (slashIdx < 0) {
                    return normalizeToPlatform(home);
                } else {
                    return normalizeToPlatform(home + afterHome.substring(slashIdx));
                }
            }
        }

        // /tmp → system temp directory
        if (path.equals("/tmp") || path.startsWith("/tmp/")) {
            String tempDir = PlatformUtils.getTempDirectory();
            if (!tempDir.isEmpty()) {
                String remainder = path.length() > "/tmp".length() ? path.substring("/tmp".length()) : "";
                return normalizeToPlatform(tempDir + remainder);
            }
        }

        return path;
    }

    // ==================== Path Length Checks ====================

    /**
     * Check whether a path length is safe (primarily for Windows).
     *
     * @param path the path to check
     * @return PathCheckResult containing the check result and recommendations
     */
    public static PathCheckResult checkPathLength(String path) {
        if (path == null || path.isEmpty()) {
            return PathCheckResult.ok();
        }

        // No check needed on non-Windows platforms
        if (!PlatformUtils.isWindows()) {
            return PathCheckResult.ok();
        }

        int pathLength = path.length();

        if (pathLength >= WINDOWS_MAX_PATH) {
            return PathCheckResult.error(
                "Project path is too long (" + pathLength + " characters)\n" +
                "Windows limits path length to 260 characters.\n" +
                "Suggestion: move the project to a shorter path, e.g. D:\\projects\\",
                path,
                pathLength
            );
        }

        if (pathLength >= SAFE_PATH_LENGTH) {
            return PathCheckResult.warning(
                "Project path is relatively long (" + pathLength + " characters)\n" +
                "It may exceed the Windows path limit when creating deeply nested files.",
                path,
                pathLength
            );
        }

        return PathCheckResult.ok(path, pathLength);
    }

    // ==================== Temporary Directory Detection ====================

    /**
     * Get all possible temporary directory paths.
     *
     * @return list of temporary directory paths
     */
    public static List<String> getTempPaths() {
        Set<String> paths = new HashSet<>();

        // Primary temp directory from PlatformUtils (handles TEMP/TMP/TMPDIR/java.io.tmpdir)
        String primaryTemp = PlatformUtils.getTempDirectory();
        if (!primaryTemp.isEmpty()) {
            paths.add(normalizeToUnix(primaryTemp).toLowerCase());
        }

        // java.io.tmpdir as additional fallback
        String javaTmpDir = System.getProperty("java.io.tmpdir");
        if (javaTmpDir != null && !javaTmpDir.isEmpty()) {
            paths.add(normalizeToUnix(javaTmpDir).toLowerCase());
        }

        if (PlatformUtils.isWindows()) {
            String localAppData = PlatformUtils.getEnvIgnoreCase("LOCALAPPDATA");
            if (localAppData != null && !localAppData.isEmpty()) {
                paths.add(normalizeToUnix(localAppData + "\\Temp").toLowerCase());
            }
        } else {
            paths.add("/tmp");
            paths.add("/var/tmp");
            paths.add("/private/tmp"); // macOS
        }

        return new ArrayList<>(paths);
    }

    /**
     * Check whether a given path is a temporary directory.
     *
     * @param path the path to check
     * @return true if the path is a temp directory or is located within one
     */
    public static boolean isTempDirectory(String path) {
        if (path == null || path.isEmpty()) {
            return false;
        }

        String normalizedPath = normalizeToUnix(path).toLowerCase();
        List<String> tempPaths = getTempPaths();

        for (String tempPath : tempPaths) {
            if (tempPath != null && normalizedPath.startsWith(tempPath)) {
                return true;
            }
        }

        return false;
    }

    /**
     * Paths that must never be used as the agent working directory / agy
     * {@code workspaceDirs}. Observed failures used plugin install dirs, the
     * embedded ai-bridge tree, or {@code ~/.gemini} when cwd fell through.
     *
     * @param path candidate working directory
     * @return true when the path is empty or clearly not a user project root
     */
    public static boolean isUnsafeWorkingDirectory(String path) {
        if (path == null || path.trim().isEmpty()) {
            return true;
        }
        String normalized = normalizeToUnix(path.trim()).toLowerCase();
        if (normalized.isEmpty() || "undefined".equals(normalized) || "null".equals(normalized)) {
            return true;
        }
        // JetBrains plugin install / config trees (embedded ai-bridge lives here).
        if (normalized.contains("/application support/jetbrains/")) {
            return true;
        }
        if (normalized.contains("/plugins/idea-claude-code-gui")) {
            return true;
        }
        if (normalized.contains("/.local/share/jetbrains/")) {
            return true;
        }
        // Standalone ai-bridge checkout used as process.cwd() by the daemon.
        if (normalized.endsWith("/ai-bridge") || normalized.contains("/ai-bridge/")) {
            return true;
        }
        // Antigravity CLI default home — was seen as workspaceDirs=[~/.gemini].
        if (normalized.endsWith("/.gemini") || normalized.contains("/.gemini/")) {
            return true;
        }
        if (normalized.endsWith("/.gemini/antigravity-cli")
                || normalized.contains("/.gemini/antigravity-cli/")) {
            return true;
        }
        return false;
    }

    /**
     * Prefer {@code requested} when it is a safe existing directory; otherwise
     * fall back to {@code projectBasePath} when that is safe. Returns the first
     * usable path, or {@code null} when neither is acceptable.
     *
     * <p>Unlike {@link #guardWorkingDirectory(String, String)} (which clamps a
     * daemon cwd inside the project), this selects an acceptable location at all:
     * agy uses process cwd as {@code workspaceDirs}, so spawning from a plugin /
     * ai-bridge / {@code ~/.gemini} tree must fall through, not clamp.
     */
    public static String selectSafeWorkingDirectory(String requested, String projectBasePath) {
        String primary = firstSafeExistingDirectory(requested);
        if (primary != null) {
            return primary;
        }
        String fallback = firstSafeExistingDirectory(projectBasePath);
        if (fallback != null) {
            return fallback;
        }
        return null;
    }

    private static String firstSafeExistingDirectory(String path) {
        if (path == null || path.trim().isEmpty() || isUnsafeWorkingDirectory(path)) {
            return null;
        }
        String normalized = normalizeAbsolute(path.trim());
        if (normalized == null || normalized.isEmpty() || isUnsafeWorkingDirectory(normalized)) {
            return null;
        }
        File dir = new File(normalized);
        if (!dir.isDirectory()) {
            return null;
        }
        return normalized;
    }

    // ==================== Path Validation ====================

    /**
     * Check whether a path is writable.
     *
     * @param path the path to verify
     * @return true if the path exists and is writable
     */
    public static boolean isWritable(String path) {
        if (path == null || path.isEmpty()) {
            return false;
        }
        File file = new File(path);
        return file.exists() && file.canWrite();
    }

    /**
     * Check whether a path exists.
     *
     * @param path the path to verify
     * @return true if the path exists
     */
    public static boolean exists(String path) {
        if (path == null || path.isEmpty()) {
            return false;
        }
        return new File(path).exists();
    }

    /**
     * Get the parent directory of a path.
     *
     * @param path the file or directory path
     * @return the parent directory path, or null if unavailable
     */
    public static String getParentPath(String path) {
        if (path == null || path.isEmpty()) {
            return null;
        }
        File file = new File(path);
        File parent = file.getParentFile();
        return parent != null ? parent.getAbsolutePath() : null;
    }

    /**
     * Join two path segments.
     *
     * @param basePath     the base path
     * @param relativePath the relative path to append
     * @return the combined path
     */
    public static String joinPath(String basePath, String relativePath) {
        if (basePath == null || basePath.isEmpty()) {
            return relativePath;
        }
        if (relativePath == null || relativePath.isEmpty()) {
            return basePath;
        }
        return new File(basePath, relativePath).getAbsolutePath();
    }
}

File: ai-bridge/utils/path-utils.js
/**
 * 路径处理工具模块
 * 负责路径规范化、临时目录检测、工作目录选择
 */

import fs from 'fs';
import { resolve, join } from 'path';
import { homedir, tmpdir } from 'os';

// 缓存真实的用户目录路径，避免重复计算
let cachedRealHomeDir = null;

/**
 * 获取真实的用户目录路径.
 * 解决 Windows 上用户目录被移动或使用符号链接/Junction 的问题。
 * 使用 fs.realpathSync 获取物理路径，确保与文件系统实际路径一致。
 * @returns {string} 真实的用户目录路径
 */
export function getRealHomeDir() {
  if (cachedRealHomeDir) {
    return cachedRealHomeDir;
  }

  const rawHome = homedir();
  try {
    // 使用 realpathSync 获取真实的物理路径，解决符号链接/Junction 问题
    cachedRealHomeDir = fs.realpathSync(rawHome);
  } catch {
    // 如果 realpath 失败，回退到原始路径
    console.warn('[path-utils] Failed to resolve real home path, using raw path:', rawHome);
    cachedRealHomeDir = rawHome;
  }

  return cachedRealHomeDir;
}

/**
 * 获取 .codemoss 配置目录路径.
 * @returns {string} ~/.codemoss 目录路径
 */
export function getCodemossDir() {
  return join(getRealHomeDir(), '.codemoss');
}

/**
 * 获取 .claude 配置目录路径.
 * @returns {string} ~/.claude 目录路径
 */
export function getClaudeDir() {
  return join(getRealHomeDir(), '.claude');
}

/**
 * 获取系统临时目录前缀列表
 * 支持 Windows、macOS 和 Linux
 */
export function getTempPathPrefixes() {
  const prefixes = [];

  // 1. 使用 os.tmpdir() 获取系统临时目录
  const systemTempDir = tmpdir();
  if (systemTempDir) {
    prefixes.push(normalizePathForComparison(systemTempDir));
  }

  // 2. Windows 特定环境变量
  if (process.platform === 'win32') {
    const winTempVars = ['TEMP', 'TMP', 'LOCALAPPDATA'];
    for (const varName of winTempVars) {
      const value = process.env[varName];
      if (value) {
        prefixes.push(normalizePathForComparison(value));
        // Windows Temp 通常在 LOCALAPPDATA\Temp
        if (varName === 'LOCALAPPDATA') {
          prefixes.push(normalizePathForComparison(join(value, 'Temp')));
        }
      }
    }
    // Windows 默认临时路径
    prefixes.push('c:\\windows\\temp');
    prefixes.push('c:\\temp');
  } else {
    // Unix/macOS 临时路径前缀
    prefixes.push('/tmp');
    prefixes.push('/var/tmp');
    prefixes.push('/private/tmp');

    // 环境变量
    if (process.env.TMPDIR) {
      prefixes.push(normalizePathForComparison(process.env.TMPDIR));
    }
  }

  // 去重
  return [...new Set(prefixes)];
}

/**
 * 规范化路径用于比较
 * Windows: 转小写，使用正斜杠
 */
export function normalizePathForComparison(pathValue) {
  if (!pathValue) return '';
  let normalized = pathValue.replace(/\\/g, '/');
  if (process.platform === 'win32') {
    normalized = normalized.toLowerCase();
  }
  return normalized;
}

/**
 * 清理路径
 * @param {string} candidate - 候选路径
 * @returns {string|null} 规范化后的路径或 null
 */
export function sanitizePath(candidate) {
  if (!candidate || typeof candidate !== 'string' || candidate.trim() === '') {
    return null;
  }
  try {
    return resolve(candidate.trim());
  } catch {
    return null;
  }
}

/**
 * 检查路径是否为临时目录
 * @param {string} pathValue - 路径
 * @returns {boolean}
 */
export function isTempDirectory(pathValue) {
  if (!pathValue) return false;

  const normalizedPath = normalizePathForComparison(pathValue);
  const tempPrefixes = getTempPathPrefixes();

  return tempPrefixes.some(tempPath => {
    if (!tempPath) return false;
    return normalizedPath.startsWith(tempPath) ||
           normalizedPath === tempPath;
  });
}

/**
 * 智能选择工作目录
 * @param {string} requestedCwd - 请求的工作目录
 * @returns {string} 选定的工作目录
 */
export function selectWorkingDirectory(requestedCwd) {
  const candidates = [];

  const envProjectPath = process.env.IDEA_PROJECT_PATH || process.env.PROJECT_PATH;

  if (requestedCwd && requestedCwd !== 'undefined' && requestedCwd !== 'null') {
    candidates.push(requestedCwd);
  }
  if (envProjectPath) {
    candidates.push(envProjectPath);
  }

  candidates.push(process.cwd());
  candidates.push(getRealHomeDir());

  console.log('[DEBUG] selectWorkingDirectory candidates:', JSON.stringify(candidates));

  for (const candidate of candidates) {
    const normalized = sanitizePath(candidate);
    if (!normalized) continue;

    if (isTempDirectory(normalized) && envProjectPath) {
      console.log('[DEBUG] Skipping temp directory candidate:', normalized);
      continue;
    }

    try {
      const stats = fs.statSync(normalized);
      if (stats.isDirectory()) {
        console.log('[DEBUG] selectWorkingDirectory resolved:', normalized);
        return normalized;
      }
    } catch {
      // Ignore invalid candidates
      console.log('[DEBUG] Candidate is invalid:', normalized);
    }
  }

  console.log('[DEBUG] selectWorkingDirectory fallback triggered');
  return envProjectPath || getRealHomeDir();
}
